src/pages/posts/comparing-rust-smart-contract-storage-approaches.md

```diff
@@ -0,0 +1,515 @@
+# Simple data
+
+As a baseline, let's see the most common approaches to storing and incrementing a number.
+
+## NEAR
+
+```rust
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::near_bindgen;
+
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+pub struct Contract {
+    counter: i8,
+}
+
+#[near_bindgen]
+impl Contract {
+    pub fn get_num(&self) -> i8 {
+        self.counter
+    }
+
+    pub fn increment(&mut self) {
+        self.counter += 1;
+    }
+}
+```
+
+If you're used to Rust, then this all looks pretty familiar. You can tell that the macros are doing all the heavy lifting.
+
+You need to add the `#[near_bindgen]` macro to a `struct`, which will make that struct the main data structure of the contract. This means all of its keys will be stored in the on-chain key-value store.
+
+NEAR also makes you import some [Borsh](https://borsh.io/) stuff to explicitly declare the encoding/decoding format of the main data structure. You can [read more about this](https://raen.dev/guide/counter/intro.html#borsh) if you're unfamiliar with the specifics of encoding and decoding. But I don't know why `near-sdk-rs` makes you be explicit in this way—every contract I've ever seen encodes and decodes with Borsh.
+
+Anyhow, the Borsh encoding means that the actual stored key will not by `counter`, as shown in the code. It will be something much smaller, like a single byte.
+
+Then you declare `#[near_bindgen]` on an implementation, `impl`, which adds functions to your main struct. All of the public (`pub`) functions in that `impl` will be exported from the contract as functions.
+
+If you want to play with a more complete example of this, check out the [RAEN Guide](https://raen.dev/guide/counter/intro.html).
+
+## Soroban
+
+Soroban is still in its early days, and has the simplest and most explicit approach to accessing the underlying key-value store:
+
+```rust
+#![no_std]
+use soroban_sdk::{contractimpl, symbol, Env, Symbol};
+
+const COUNTER: Symbol = symbol!("COUNTER");
+
+pub struct IncrementContract;
+
+#[contractimpl]
+impl IncrementContract {
+    pub fn get_num(env: Env) -> i32 {
+        env.data()
+            .get(COUNTER) // Returns an Option (Some or None) wrapping a Result (Ok or Err).
+            .unwrap_or(Ok(0)) // Unwrap the Option. If None, no value set. Default to 0.
+            .unwrap() // Unwrap the Result. If Err, COUNTER is not i32. Panic.
+    }
+
+    pub fn increment(env: Env) {
+        let count: i32 = env.data().get(COUNTER).unwrap_or(Ok(0)).unwrap();
+        env.data().set(COUNTER, count + 1);
+    }
+}
+```
+
+Soroban requires skipping the standard library (`#![no_std]`), which means you can't use [String](https://doc.rust-lang.org/std/string/struct.String.html), only [slices](https://doc.rust-lang.org/book/ch04-03-slices.html). It provides the `symbol!` macro to set the key that is used in storage, and the actual full `COUNTER` string will be the actual storage key.
```

Comment:
The [symbol type](https://soroban.stellar.org/docs/learn/built-in-types#symbol-symbol) is not exactly a string

> Symbols are small efficient strings up to 10 characters in length and limited to a-z A-Z 0-9 _ that are encoded into 64-bit integers.

(although there's a [discussion](https://github.com/stellar/rs-soroban-env/issues/463) on changing this)

Reply:
What is the key in the key-value store? Using `soroban read` makes it seem like it is `COUNTER`, but maybe it's doing a conversion?

Reply:
The key is a Symbol, but the Soroban read command does some coercion of types, like a string it will convert into a Symbol.

However, a Symbol is actually a string, depending on context. Inside contracts it is an efficiently stored string, that is stored as a u64. Outside contracts it is ASCII-compatible 10 bytes.

Reply:
You can use `&str` in soroban contracts if you'd like, although the use cases are unclear to me.

Reply:
Also, the only reason you can't use String is because there's no allocator, but I think you could provide your own allocator and then use it. The SDK just doesn't setup an allocator for you. This is something I'd like to explore at some point, but admittedly it is down the list.

Reply:
I've also wanted to try https://crates.io/crates/compact_str

Reply:
@graydon are you still planning to explore a string type for Soroban and a constant slice type that's going to do some magic for longer values? I think yes.

Reply:
Ok yeah found the issues:
- https://github.com/stellar/rs-soroban-env/issues/463
- https://github.com/stellar/rs-soroban-env/issues/584

Reply:
@leighmcculloch thanks for all the links! I'm wondering what the actual bytes in the key-value store would be, in this case. Like, the key, in the key-value store. `COUNTER` encoded how? UTF-8? Is it some representation of that actual set of characters, or is it condensed down to one byte or something?

Reply:
> I'm wondering what the actual bytes in the key-value store would be

Symbols are stored efficiently inside a u64 (it fits into 60 bites actually) inside the Soroban contract runtime, but outside of it they're stored fully expanded and not particularly compact. Here's a rundown:

All keys and values are stored as an `ScVal`. The XDR definition for that is here: https://github.com/stellar/stellar-xdr/blob/026c9cd074bdb28ddde8ee52f2a4502d9e518a09/Stellar-contract.x#L187-L205. An `ScVal` is a discriminated union ([ref](https://www.rfc-editor.org/rfc/rfc4506#section-4.15)) which means it can contain several different types of value, and to specify which value is being stored it first stores a 4-byte discriminant (enum) to indicate what value is inside.

If a Symbol is being stored, that means it will be an `ScVal::Symbol(ScSymbol)` in the Rust code, where an `ScSymbol` is a variable length string (byte array) with a maximum length of 10. On disk this will be stored as:

```
00 00 00 05 // ScVal's discriminant, an ScValType
00 00 00 0a // ScSymbo's length prefix, a 4 byte length, here lets say it is 10
xx xx xx xx xx xx xx xx xx xx  // The symbol string bytes 
00 00 // 2 byte padding, as xdr requires all types to be 4-byte aligned
```

To double check this:

```
$ echo -e -n "\x00\x00\x00\x05\x00\x00\x00\x0a\x68\x65\x6c\x6c\x6f\x20\x77\x72\x6c\x64\x00\x00" > val.xdr

$ xxd val.xdr
00000000: 0000 0005 0000 000a 6865 6c6c 6f20 7772  ........hello wr
00000010: 6c64 0000                                ld..

$ cargo install --locked stellar-xdr --features cli

$ stellar-xdr +next decode --type ScVal --input single val.xdr
{"symbol":"hello wrld"}
```

Reply:
Thanks for all the info, @leighmcculloch! Here's my proposed rewrite of the paragraph in question; how's it sound?

```suggestion
Soroban requires skipping the standard library (`#![no_std]`), which means you can't use [String](https://doc.rust-lang.org/std/string/struct.String.html), only [slices](https://doc.rust-lang.org/book/ch04-03-slices.html). The [`symbol!` macro](https://docs.rs/soroban-sdk/0.3.1/soroban_sdk/macro.symbol.html) handles converting user-provided, ASCII-compatible strings of max-length 10 into an efficient 64-bit runtime [value type](https://soroban.stellar.org/docs/learn/environment-concepts#values-and-types). When persisted to the underlying key-value store, this `COUNTER` symbol gets encoded with [XDR](https://www.rfc-editor.org/rfc/rfc4506), which includes one byte each for the letters in `COUNTER`, plus some [extra encoding information](https://github.com/AhaLabs/ahalabs.github.io/pull/11/files#r1039065050) so that every key in the key-value store ends up using 20 bytes.
```